test/parallel/test-process-codegen-from-string.js

```diff
@@ -0,0 +1,24 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+
+const item = { foo: 0 };
+eval('++item.foo');
+
+assert.strictEqual(item.foo, 1);
+
+process.on('codeGenerationFromString', common.mustCall((code) => {
+  assert.strictEqual(code, 'item.foo++');
+}));
```

Comment:
As @vdeturckheim pointed out in a comment on this PR, there's something not working as intend in this callback:

If an error is thrown inside of this callback, it doesn't always get picked up. In this example code, the callback will get called, but the error will be silenced:

```js
process.on('codeGenerationFromString', (code) => {
  console.log('Hello from callback!')
  throw new Error('boom!!!') // error is silenced
})

eval('1+1')
```

However, this code will _not_ silence the error:

```js
process.on('codeGenerationFromString', (code) => {
  console.log('Hello from callback!')
  throw new Error('boom!!!') // error is not silenced
})

console.log('result of eval:', eval('1+1'))
```

I found that queing a microtask after the call to eval is also sufficient:

```js
process.on('codeGenerationFromString', (code) => {
  console.log('Hello from callback!')
  throw new Error('boom!!!') // error is not silenced
})

eval('1+1')
queueMicrotask(() => {})
```

However, `process.exit(() => {})` is not.

On top of this, if the listener is removed before the microtask is queued, the error is once again swallowed:

```js
process.on('codeGenerationFromString', (code) => {
  console.log('Hello from callback!')
  throw new Error('boom!!!') // error is silenced
})

eval('1+1')

process.removeAllListeners('codeGenerationFromString');
queueMicrotask(() => {})
```

Reply:
My initial guess was that the error was handled asynchronously and hence if the program ended or the listener was removed before Node.js had a chance to handle the error then it would be lost. However, if I swap the `removeAllListeners` and the `queueMicrotask` calls in the last code-example above, the error is actually retained, which I don't think it should according to that theory. So the very act of queueing the microtask - not actually waiting for it to be executed - seems to be enough for it to retain the error before the listener is removed.

Reply:
New discovery: If I add a line of invalid code after removing the listeners, I don't get any error complaining about that line either. So it looks like it doesn't get to that line at all:

```js
process.on('codeGenerationFromString', (code) => {
  console.log('Hello from callback!')
  throw new Error('boom!!!') // error is silenced
})

eval('1+1')

process.removeAllListeners('codeGenerationFromString')
bad_code
```

Reply:
I tried to compile Node.js with debug mode enabled (`./configure --debug && make -j8`) and ran it via lldb with breakpoints set to `_exit` and `exit`. And it looked like the program just ended normally like a regular program (which also explains the zero exit code).

Reply:
Ok, with a lot of help from @addaleax I finally understand why the error was "silenced":

If V8 doesn't expect a callback given to it to call into JS-land, any JS exception thrown will not be properly handled. So an explicit try-catch has to be added to the C++ code that's calling into JS to ensure that the exception is handled. We have the same problem here:

https://github.com/nodejs/node/blob/fa7cdd6fc9dffa8139f9350f54959b01bf7a1151/src/node_task_queue.cc#L153-L159

I've updated this PR accordingly.

However, this unfortunately means that Node.js will not exit with an uncaught exception and so this isn't easily tested in our tests (except if we check the output to `stderr` I guess). And it also can be quite confusing for users if they except Node.js to crash.

Reply:
Would `process.on('uncaughtexception')` still fire?

Reply:
@vdeturckheim No it will not. The exception is now technically handled